src/libcore/fmt/builders.rs

```diff
@@ -112,11 +131,16 @@ impl<'a, 'b: 'a> DebugStruct<'a, 'b> {
             };
 
             if self.is_pretty() {
+                use fmt::Write;
                 let mut writer = PadAdapter::new(self.fmt);
-                fmt::write(&mut writer,
-                           format_args!("{}\n{}: {:#?}", prefix, name, value))
+                writer.write_str(prefix)?;
+                writer.write_str("\n")?;
+                writer.write_str(name)?;
+                writer.write_str(": ")?;
+                value.fmt(&mut writer.as_formatter())
```

Comment:
What's the difference between using this and `value.fmt(writer.fmt)` here?

Reply:
That would break the indentation and newline insertion that `PadAdapter` is there to do, which is particularly useful for deeply nested structures.

Reply:
I see! I thinking we could have written this without privileged access to formatter internals, but this way has less indirections and in debug struct it really benefits everyone, so I think it serves its purpose.

Reply:
I tried making this a method of `Formatter` instead, but couldn’t borrow `&mut writer` and `&mut writer.fmt` at the same time.

Reply:
Alright, I found a way to make it work (in a new commit) but it’s a bit messy with up to three named lifetimes in the same item. I’m not sure that it’s better than before. What do you think?

Reply:
I think it looks quite neat. I'd pinpoint two places to place comments and it's good for me: 1. What's the slot argument?, 2. One sentence explaining the wrap_buf method. (The comment "We want to change this" is a sign that a prior brief comment is missing IMO)